src/MongoDB.Driver.Core/Core/Misc/ServerVersion.cs

```diff
@@ -0,0 +1,505 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace MongoDB.Driver.Core.Misc
+{
+    internal class ServerVersion : IEquatable<ServerVersion>, IComparable<ServerVersion>
+    {
+        #region static
+        private static void LookForPreReleaseNumericSuffix(string preRelease, out string preReleasePrefix, out int? preReleaseNumericSuffix)
+        {
+            var pattern = @"^(?<prefix>.*[^\d])(?<numericSuffix>\d+)$";
+            var match = Regex.Match(preRelease, pattern);
+            if (match.Success)
+            {
+                preReleasePrefix = match.Groups["prefix"].Value;
+                preReleaseNumericSuffix = int.Parse(match.Groups["numericSuffix"].Value);
+            }
+            else
+            {
+                preReleasePrefix = preRelease;
+                preReleaseNumericSuffix = null;
+            }
+        }
+
+        private static bool TryParseInternalPrelease(string preReleaseIn, out string preReleaseOut, out int? commitsAfterRelease, out string commitHash)
+        {
+            if (preReleaseIn != null)
+            {
+                var internalBuildPattern = @"^(?<preRelease>[A-Za-z]+[0-9]*)?-?(?<commitsAfterRelease>\d+)-g(?<commitHash>[0-9a-fA-F]{4,40})$";
```

Comment:
I think the pattern is a little to restrictive. I suggest:

    var internalBuildPattern = @"^((?<preRelease>.+)-)?(?<commitsAfterRelease>\d+)-g(?<commitHash>[0-9a-fA-F]{4,40})$";